buildtest/modules/util.py

```diff
@@ -1,48 +1,156 @@
-import subprocess
+from buildtest.tools.system import BuildTestCommand
 
 
 def get_all_collections():
-    """Return all module collection """
+    """Get all Lmod user collections that is retrieved by running ``module -t savelist``.
+
+     :return: Return all module collections
+     :rtype: list
+     """
+
     collections = "module -t savelist"
-    ret = subprocess.Popen(
-        collections, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-    )
-    out = ret.communicate()[1]
-    tmp = out.decode("utf-8").split()
+    cmd = BuildTestCommand()
+    cmd.execute(collections)
+    # The output is captured in STDERR stream.
+    out = cmd.get_error().split()
 
-    return tmp
+    return out
 
 
 class ModuleCollection:
+    """Class declaration of ModuleCollection."""
+
     def __init__(self, collection):
+        """Initializer method of ModuleCollection class.
```

Comment:
Can you walk me through the difference in just a module / module collection? I'm guessing it's an organized group of modules - is this something created by buildtest, or associated with Lmod? How does this link up to a user request to run just one recipe file:

```bash
buildtest build -c /path/to/mpi.yml
```
Versus an entire folder?

```bash
buildtest build -d /path/to/collection
```
Couldn't it be the case that there is a recipe (defined in a collection) that you would target the individual file for? Or maybe what I'm asking - what's the distinction between a collection and just a folder / subset of recipes, and given that we can organize with folders, why do we need collections?

Reply:
modules and module collection are means to load the modules which are going to be used when defining the YAML schema. Currently the buildtest allowed few ways to load modules, two of them are the module collection and buildtest own module collection. 

See https://buildtest.readthedocs.io/en/latest/build_subcommand/module_collections.html

The schema needs to be changed, before user must specify ``buildtest build -c <config> -co <collection>`` or ``buildtest build -c <config> -mc <buildtest-collection-ID>`` to allow loading modules. If none are specified then active modules are loaded.

The idea was having a Module API which would allow a mechanism to interact with module system. 

If you are new to Lmod user collection check out https://lmod.readthedocs.io/en/latest/010_user.html#user-collections. 

Idea is to allow buildtest to write tests where user can specify how to load modules. Either user can create a Lmod user collection. Or use the buildtest module collection see https://buildtest.readthedocs.io/en/latest/module_operation.html#module-collection-operation-buildtest-module-collection which was implemented to allow an arbitrary size of module collections. The idea was to use collection ID instead of unique names since user who use collection names need to write unique names and it will cause issue when there are alot of user collection.

The ``ModuleCollection`` class is provided to interact with module collection. The ``Module`` class is the main class that user can interact with modules.


I have few examples of the YAML schema in this wiki https://github.com/HPC-buildtest/buildtest-framework/wiki/Yaml-Schema-Draft that i was thinkin about the new schema. This is not implemented.

Hope that answers few questions and maybe leads you to more questions :)

Reply:
Interesting - my intuition is that the typical sysadmin would store module "collections" on the filesystem in folders, so this seems like a feature that would be afforded by just having folders instead, and then possibly using an include / exclude filter. 

But actually, since it's integrated with Lmod, this is probably something that is used and should be represented here. I can get a better sense when I am able to create my own little collection, and set of modules! I'll also ask my group about how they use collections. Either way, I think we should be careful with the UI, because there are now several different ways to select some subset of modules (and too many might be confusing for the UI):

 - select a config directly with -c / --config (collections not relevant)
 - select a directory with -d (with some include/exclude filters) - are collections relevant here? They probably shouldn't be.
 - select one or more collections
 - don't select anything, but discover all defined in the config.

Anyway, I am +1 for having collections given that they are a big part of Lmod - I'll need to refresh my learning about how they work to best discuss the implementation.

Reply:
Lmod collection has been around for quite some time. Load a bunch of modules and save them into a collection, then purge your modules and you can restore all your modules using ``module restore <collection>``. 

I think we may be confusing some terminology, the ``--config`` and supposedly ``-d`` (not implemented) are options to deal with which YAML file to process. Inside the YAML you can specify key and one of them could be how to load modules. 

In the backend, buildtest needs to call the Module API to load the appropriate module. 

For example this example is showing that we can use a collection name ``GCC`` to restore the module. Assuming that user has this collection saved. The collections are saved in $HOME/.lmod.d/ 

```
# module restore example for C
hello_ex1:
  type: singlesource
  description: "C++ hello world example"
  compiler: gnu
  source: hello.cpp
  scheduler: local
  # module collection to restore
  moduleload:
    collection:
      name:
       - GCC
  cxxflags:
    - "-O2"
```

There can be other examples such as ones below

```
# module load example with one version
hello_ex1:
  type: singlesource
  description: "C hello world example1"
  compiler: gnu
  source: hello.c
  scheduler: local
  moduleload:
    key:
      name: GCC
      version: ["6.4.0"]
  cflags:
    - "-O1"

# single test with all version
hello_ex2:
  type: singlesource
  description: "C hello world example2"
  compiler: gnu
  source: hello.c
  scheduler: local
  moduleload:
    key:
      name: GCC
      version: ["*"]
  cflags:
    - "-O1"
```


The second example where ``version: ["*"]`` is doing a test against all GCC modules. This is not implemented in the ModuleAPI and it is pretty simple to get this since I have most of the methods implemented already.